File: Algorithms/Greedy_algorithms/job_sequencing.c
// Given a set of jobs with their deadlines and their profits, find the maximum profit that can be earned
// by scheduling the jobs without missing any deadlines

#include <stdio.h>

struct Job {
    char id;
    int deadline;
    int profit;
};

void jobSequencing(struct Job jobs[], int n) {
    int maxDeadline = 0;
    for (int i = 0; i < n; i++) {
        if (jobs[i].deadline > maxDeadline) {
            maxDeadline = jobs[i].deadline;
        }
    }

    char sequence[maxDeadline];
    for (int i = 0; i < maxDeadline; i++) {
        sequence[i] = ' ';
    }

    for (int i = 0; i < n; i++) {
        for (int j = jobs[i].deadline - 1; j >= 0; j--) {
            if (sequence[j] == ' ') {
                sequence[j] = jobs[i].id;
                break;
            }
        }
    }

    int totalProfit = 0;
    printf("Job sequence: ");
    for (int i = 0; i < maxDeadline; i++) {
        if (sequence[i] != ' ') {
            printf("%c ", sequence[i]);
            totalProfit += jobs[i].profit;
        }
    }
    printf("\nTotal profit: %d\n", totalProfit);
}

int main() {
    struct Job jobs[] = { {'a', 2, 100}, {'b', 1, 19}, {'c', 2, 27}, {'d', 1, 25}, {'e', 3, 15} };
    int n = sizeof(jobs) / sizeof(jobs[0]);
    jobSequencing(jobs, n);
    printf("\n");
    return 0;
}

File: Algorithms/Greedy_algorithms/job_sequencing.py
# Given a set of jobs with their deadlines and their profits, find the maximum profit that can be earned
# by scheduling the jobs without missing any deadlines

class Job:
    def __init__(self, id, deadline, profit):
        self.id = id
        self.deadline = deadline
        self.profit = profit

def job_sequencing(jobs):
    max_deadline = max(jobs, key=lambda x: x.deadline).deadline

    sequence = [' '] * max_deadline

    for job in jobs:
        for j in range(job.deadline - 1, -1, -1):
            if sequence[j] == ' ':
                sequence[j] = job.id
                break

    total_profit = 0
    print("Job sequence:", end=" ")
    for job_id in sequence:
        if job_id != ' ':
            print(job_id, end=" ")
            job = next((job for job in jobs if job.id == job_id), None)
            total_profit += job.profit
    print()
    print("Total profit:", total_profit)

jobs = [Job('a', 2, 100), Job('b', 1, 19), Job('c', 2, 27), Job('d', 1, 25), Job('e', 3, 15)]
job_sequencing(jobs)
